src/modules/job-manager/purge.c

```diff
@@ -0,0 +1,408 @@
+                                       0,
+                                       "{s:O}",
+                                       "jobs", jobs)))
+        return -1;
+    flux_future_destroy (f);
+    return 0;
+}
+
+/* Return true if candidate job is eligible for purging based on
+ * provided limits.  A job is purgeable if either limit is exceeded.
+ */
+static bool purge_eligible (double age,
+                            double age_limit,
+                            int num,
+                            int num_limit)
+{
+    if (age_limit != INACTIVE_AGE_UNLIMITED && age > age_limit)
+        return true;
+    if (num_limit != INACTIVE_NUM_UNLIMITED && num > num_limit)
+        return true;
+    return false;
+}
+
+/* Send a KVS commit containing unlinks for one or more inactive jobs.
+ * Return future if successful, with 'count' added to aux hash.
+ * Return NULL on failure with errno set.
+ * N.B. Failure with errno=ENODATA just means there were no eligible jobs.
+ */
+static flux_future_t *purge_inactive_jobs (struct purge *purge,
+                                           double age_limit,
+                                           int num_limit,
+                                           int max_purge_count)
+{
+    double now = flux_reactor_now (flux_get_reactor (purge->ctx->h));
+    struct job *job;
+    flux_kvs_txn_t *txn;
+    json_t *jobs = NULL;
+    char key[64];
+    flux_future_t *f = NULL;
+    int count = 0;
+
+    if (!(txn = flux_kvs_txn_create ()))
+        return NULL;
+    if (!(jobs = json_array ())) {
+        errno = ENOMEM;
+        goto error;
+    }
+    while ((job = zlistx_first (purge->queue)) && count < max_purge_count) {
+        if (!purge_eligible (now - job->t_clean,
+                             age_limit,
+                             zlistx_size (purge->queue),
+                             num_limit))
+            break;
+        if (flux_job_kvs_key (key, sizeof (key), job->id, NULL) < 0
+            || flux_kvs_txn_unlink (txn, 0, key) < 0)
+            goto error;
+        json_t *o = json_integer (job->id);
+        if (!o || json_array_append_new (jobs, o)) {
+            json_decref (o);
+            errno = ENOMEM;
+            goto error;
+        }
+        (void)zlistx_delete (purge->queue, job->handle);
+        job->handle = NULL;
+        zhashx_delete (purge->ctx->inactive_jobs, &job->id);
+        count++;
+    }
+    if (count == 0) {
+        errno = ENODATA;
+        goto error;
+    }
+    if (!(f = flux_kvs_commit (purge->ctx->h, NULL, 0, txn))
+        || flux_future_aux_set (f, "count", int2ptr (count), NULL) < 0
+        || purge_publish (purge, jobs) < 0)
+        goto error;
+    flux_kvs_txn_destroy (txn);
+    /* N.B. if kvs commit fails, jobs are still removed from the hash/list.
+     * It doesn't seem worth the effort to structure the code to avoid this
+     * due to high complexity of solution, low probability of error, and
+     * minor consequences.
+     */
```

Comment:
I didn't have a chance to fully review this PR yet, but `FLUX_TEST_VALGRIND=t ./t2809-job-purge.t` caught a leak of `jobs` on successful return here.

```
==89443== 1,168 (320 direct, 848 indirect) bytes in 8 blocks are definitely lost in loss record 21 of 29
==89443==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==89443==    by 0x4A25070: json_array (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.13.0)
==89443==    by 0xA39A93C: purge_inactive_jobs (purge.c:110)
==89443==    by 0xA39AC50: purge_request_cb (purge.c:283)
==89443==    by 0x487B8B7: call_handler (msg_handler.c:344)
==89443==    by 0x487BC49: dispatch_message (msg_handler.c:380)
==89443==    by 0x487BC49: handle_cb (msg_handler.c:481)
==89443==    by 0x48AACE2: ev_invoke_pending (ev.c:3770)
==89443==    by 0x48AE02F: ev_run (ev.c:4190)
==89443==    by 0x48AE02F: ev_run (ev.c:4021)
==89443==    by 0x487A9FE: flux_reactor_run (reactor.c:128)
==89443==    by 0xA392C11: mod_main (job-manager.c:220)
==89443==    by 0x119B20: module_thread (module.c:183)
==89443==    by 0x4AC4B42: start_thread (pthread_create.c:442)
```

Reply:
Ah, most likely that was in the code I just added for updating job-list.  Thank you for catching that!

Reply:
Fixed (squashed into 39664788f18df39be3c279d2c49c32d147a40126)